File: lib/common/src/Arguments.cpp
/*
 *  This file is part of OpenStaticAnalyzer.
 *
 *  Copyright (c) 2004-2018 Department of Software Engineering - University of Szeged
 *
 *  Licensed under Version 1.2 of the EUPL (the "Licence");
 *
 *  You may not use this work except in compliance with the Licence.
 *
 *  You may obtain a copy of the Licence in the LICENSE file or at:
 *
 *  https://joinup.ec.europa.eu/software/page/eupl
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the Licence is distributed on an "AS IS" basis,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the Licence for the specific language governing permissions and
 *  limitations under the Licence.
 */

#include <cstdio>
#include <cstring>
#include <cstdlib>
#include <string>
#include <list>
#include <fstream>
#include <common/inc/Arguments.h>
#include <common/inc/WriteMessage.h>
#include "ErrorCodes.h"
#include <common/inc/messages.h>

#if defined(__unix__)
#define CL_PAR_PLUS           '\''                // 'one parameter'
#elif  defined(_WIN32)
#define CL_PAR_PLUS           '"'                 // 'one parameter'
#endif

using namespace std;

#define MAX_BUFFER 4096

namespace common {

  static int countQuotationMarks(const char *s)
  {
    int counter = 0;
    const char *c = s;
    bool bs = false;
    while (*c) {
      switch (*c) {
        case '\\':
          bs = !bs;  
          break;
        case CL_PAR_PLUS:
          if (!bs)
            counter++;
          bs = false;
          break;
        default:
          bs = false;
          break;
      }
      c++;
    }
    return counter;
  }

  static string removeQuotationMarks(string s)
  {
    string::size_type ib = 0, ie = s.length()-1;
    while( s[ib] == CL_PAR_PLUS ) ++ib;
    while( s[ie] == CL_PAR_PLUS ) --ie;
    if( ib || ie != s.length()-1 )
      return s.substr(ib, ie);
    return s;
  }

  void constructArguments(const string& str, int& argc, char ** &argv)
  {
    list<string> params;   
    string param="";
    int qmc = 0;
    bool bs=false;

    unsigned int i;
    for (i=0; i<str.length(); i++){
      switch(str[i]){
        case '"':
          if (!bs)
            qmc++;
          param += str[i];
          bs = false;
          break;
        case '\\':
          bs = true;
          param += str[i];
          break;
        case '\n':
        case ' ':
          if (qmc%2){
            bs=false;
            param+=str[i];
          } else {
            params.push_back(param);
            param = "";
            qmc = 0;
            bs = false;
          }
          break;
        default:
          param += str[i];
          bs = false;
      }
    }

    // creating the new argv array
    argc = (int)params.size();
    argv = (char**)malloc(sizeof(char*) * argc);
    if (!argv){
      argc = 0;
      return;
    }

    i = 0;
    for (list<string>::const_iterator it = params.begin(); it != params.end(); ++it, i++) {   
      (argv)[i] = (char*) malloc(sizeof(char)*it->length());
      strcpy((argv)[i],(char*)it->c_str());
    }
  }

  static bool processOptionFile(const Option* options, const char *optionPrefixes, void (*unrecOption)(char*), void (*fileParam)(char*), const Option* found_option, char* found_option_args[])
  {
    if (found_option->type & OT_OPTION_FILE) {   // 
      bool ret = true;
      if (found_option->process)
        ret = found_option->process(found_option, found_option_args);

      if (ret) {   // The file have to be processed
        common::WriteMsg::write(CMSG_DUINFO_OPTIONFILE, found_option_args[0]);
        FILE *f = fopen(found_option_args[0], "rt"); 
        if (!f) {
          common::WriteMsg::write(CMSG_DUINFO_FAILED_TO_OPEN, found_option_args[0]);
          return false;
        }
        // allocating buffer for the parameters
        char *buffer = (char*)malloc(MAX_BUFFER);   
        if (!buffer) {
          fclose(f);
          return false;
        }

        list<string> params;      
        string param;
        int qmc = 0;

        // reading the parameters
        while (fscanf(f, "%s", buffer) != EOF) {
          if (qmc %2)
            param += " ";

          qmc += countQuotationMarks(buffer);
          param += buffer;

          if (qmc % 2 == 0) {
            params.push_back(removeQuotationMarks(param));

            qmc = 0;
            param = "";
          }

        }
        if (qmc % 2)
          params.push_back(removeQuotationMarks(param));

        free(buffer);
        fclose(f);

        // creating the new argv array
        int newargc = (int)params.size();
        char **newargv = (char**)malloc(sizeof(char*) * newargc);
        if (!newargv)
          return false;
    
        int i = 0;
        for (list<string>::const_iterator it = params.begin(); it != params.end(); ++it, i++) {
          common::WriteMsg::write(CMSG_DUINFO_READ_PARAM, it->c_str());
          newargv[i] = (char*)it->c_str();
        }
        bool result = processArguments(newargc, newargv, options, optionPrefixes, unrecOption, fileParam);
        free(newargv);
        return result;
      }
  
    } else
      if (found_option->process)
        return found_option->process(found_option, found_option_args);
  
    return true;    
  }

  bool processArguments(int argc, char* argv[], const Option* options, const char *optionPrefixes, void (*unrecOption)(char*), void (*fileParam)(char*))
  {
    std::list<const Option*> defCallbacks;
    
    if (!options) {
      common::WriteMsg::write(CMSG_NO_OPTION_GIVEN);
      return false;
    }
  
    // for every parameter 
    for (int pnum = 0; pnum < argc; ++pnum) {   
      const Option* found_option = NULL;
    
      // check all the given options
      for (const Option* opt = options; opt->name[0]; ++opt) {
        
        bool match = false;
      
        if (opt->type & OT_PREFIX && !(opt->type & ( OT_WOS | OT_WC | OT_WE))) { // The unambiguous prefix is enough
        
          if (strstr(opt->name, argv[pnum]) == opt->name) 
            match = true;
      
        } else {   // The prefix is not enough
        
          if (strstr(argv[pnum], opt->name) == argv[pnum]) {
            size_t oplen = strlen(opt->name);
            if (
                (!argv[pnum][oplen] && ((!opt->param_num && !(opt->type & OT_OPTION_FILE)) || (opt->param_num && (opt->type & OT_WS) && !(opt->type & OT_DEFAULT)))) ||
                (!argv[pnum][oplen] && opt->param_num == 1 && (opt->type & OT_DEFAULT) && !(opt->type & OT_WS) && !(opt->type & OT_OPTION_FILE) && !(opt->type & OT_WOS)) ||
                (argv[pnum][oplen] && opt->param_num == 1 && (opt->type & OT_WOS) && !(opt->type & OT_DEFAULT)) ||
                (argv[pnum][oplen] == '=' && opt->param_num == 1 && (opt->type & OT_WE)) ||
                (argv[pnum][oplen] == ':' && opt->param_num == 1 && (opt->type & OT_WC)))
            {
              match = true;
            }
          }
        }
      
        if (match) {
          if (found_option) {
            common::WriteMsg::write(CMSG_AMBIGUOUS_PARAM);
            return false;            
          }
          found_option = opt;    
        }
      }

      // process the option parameters
      if (found_option) {
        if (found_option->param_num == 0) {         // The option doesn't have any parameter 
          if (found_option->process && !found_option->process(found_option, NULL)) {
            common::WriteMsg::write(CMSG_FAILED_PROCESSING, found_option->name);
            return false;
          }          
        } else if (found_option->param_num == 1) {  // The option has 1 parameter


          if (found_option->type & OT_PREFIX) {    

            if (pnum + 1 >= argc) { // The parameter must be separated from the option 
              common::WriteMsg::write(CMSG_FAILED_PROCESSING, found_option->name);
              return false;
            }

            if (!processOptionFile(options, optionPrefixes, unrecOption, fileParam, found_option, argv + pnum + 1)) {
              common::WriteMsg::write(CMSG_FAILED_PROCESSING, found_option->name);
              return false;
            }
            pnum++;
          }
          else if (found_option->type & OT_DEFAULT) {
            
            bool needProcess = true;
            size_t oplen = strlen(found_option->name);
            char* prm = NULL;

            if (argv[pnum][oplen] == '=')         //  -option=value (OT_WE)
              prm = argv[pnum] + oplen + 1;
            else if (argv[pnum][oplen] == ':')    //  -option:value (OT_WC)
              prm = argv[pnum] + oplen + 1;
            else {
              if (found_option->defaultProcess != NULL) {
                //if there's no parameter for the option a default value used (calling the default callback method)
                defCallbacks.push_back(found_option);
                needProcess = false;
              } else {
                common::WriteMsg::write(CMSG_NO_DEFAULT_VALUE, found_option->name);
                return false;
              }
            }

            if (needProcess) {
              if (!processOptionFile(options, optionPrefixes, unrecOption, fileParam, found_option, &prm)) {
                common::WriteMsg::write(CMSG_FAILED_PROCESSING, found_option->name);
                return false;
              }
            }

          } else {    
            size_t oplen = strlen(found_option->name);
            char* prm = NULL;

            if (argv[pnum][oplen] == '=')         //  -option=value (OT_WE)
              prm = argv[pnum] + oplen + 1;
            else if (argv[pnum][oplen] == ':')    //  -option:value (OT_WC)
              prm = argv[pnum] + oplen + 1;
            else if (argv[pnum][oplen])           //  -optionvalue (OT_WOS)
              prm = argv[pnum] + oplen;  
            else {                                // -option value (OT_WS)
              if (pnum + 1 >= argc)  {
                common::WriteMsg::write(CMSG_FAILED_PROCESSING, found_option->name);
                return false;
              }
              prm = argv[pnum+1];
              pnum++;
            }

            if (!processOptionFile(options, optionPrefixes, unrecOption, fileParam, found_option, &prm)) {
              common::WriteMsg::write(CMSG_FAILED_PROCESSING, found_option->name);
              return false;
            }          
          }        
        } else if (found_option->param_num > 1) {   // The option has more than 1 parameter
        
          if (pnum + found_option->param_num >= argc) {
            common::WriteMsg::write(CMSG_FAILED_PROCESSING, found_option->name);
            return false;
          }

          if (found_option->process && !found_option->process(found_option, argv + pnum + 1)) {
            common::WriteMsg::write(CMSG_FAILED_PROCESSING, found_option->name);
            return false;
          }
        
          pnum += found_option->param_num;
        }
      } else {  // Unrecognized option or file
        if (strchr(optionPrefixes, argv[pnum][0]))  {  // This is an unrecegnized option
          if (unrecOption) 
            unrecOption(argv[pnum]);
          else
            common::WriteMsg::write(CMSG_UNRECOGNIZED_PARAM, argv[pnum]);
        } else {  // This is not an option
          if (fileParam)
            fileParam(argv[pnum]);
        } 
      } //if(found_option) + else
    } //for (int pnum = 0; pnum < argc; ++pnum) {
    
    std::list<const Option*>::iterator it;
    for (it = defCallbacks.begin(); it != defCallbacks.end(); it++) {
       const Option* opt = *it;
       opt->defaultProcess(*it);
    }
  
    return true;
  } //processArguments

}

File: csharp/lib/Columbus/Arguments.cs
using System;
using System.Collections.Generic;
using System.Linq;
using static Columbus.Common;

namespace Columbus
{
    public static class Arguments
    {
        public static bool ProcessArguments(string[] argv, Option[] options, string optionPrefixes, DelegateOption unrecOption)
        {
            List<Option> defCallbacks = new List<Option>();
            var neededOptions = new Dictionary<int, ICollection<Option>>();

            if (options == null)
            {
                WriteMsg.WriteLine("Error: No option given", WriteMsg.MsgLevel.Error);
                return false;
            }

            foreach (var option in options)
            {
                if (option.Needed <= 0)
                    continue;

                ICollection<Option> list;
                if (neededOptions.TryGetValue(option.Needed, out list))
                    list.Add(option);
                else
                    neededOptions.Add(option.Needed, new List<Option> {option});
            }

            for (int pnum = 0; pnum < argv.Length; ++pnum)
            {
                Option? found_option = null;

                foreach (Option opt in options)
                {
                    bool match = false;
                    if (((opt.Type & ArgumentOptions.OT_PREFIX) != ArgumentOptions.OT_NONE) && (opt.Type & (ArgumentOptions.OT_WOS | ArgumentOptions.OT_WC | ArgumentOptions.OT_WE)) == ArgumentOptions.OT_NONE)
                    {
                        if (opt.Name.ToLower().IndexOf(argv[pnum].ToLower()) == 0) match = true;
                    }
                    else
                    {
                        if (argv[pnum].ToLower().IndexOf(opt.Name.ToLower()) == 0)
                        {
                            int oplen = opt.Name.Length;
                            try
                            {
                                if (
                                    (argv[pnum].Length == oplen && (((opt.NumberOfParameters == 0 && (opt.Type & ArgumentOptions.OT_OPTION_FILE) == ArgumentOptions.OT_NONE) || (opt.NumberOfParameters == 0 && (opt.Type & ArgumentOptions.OT_WS) != ArgumentOptions.OT_NONE && (opt.Type & ArgumentOptions.OT_DEFAULT) == ArgumentOptions.OT_NONE)))) ||
                                    (argv[pnum].Length == oplen && opt.NumberOfParameters == 1 && (opt.Type & ArgumentOptions.OT_DEFAULT) != ArgumentOptions.OT_NONE && (opt.Type & ArgumentOptions.OT_WS) == ArgumentOptions.OT_NONE && (opt.Type & ArgumentOptions.OT_OPTION_FILE) == ArgumentOptions.OT_NONE && (opt.Type & ArgumentOptions.OT_WOS) == ArgumentOptions.OT_NONE) ||
                                    (argv[pnum].Length != oplen && opt.NumberOfParameters == 1 && (opt.Type & ArgumentOptions.OT_WOS) != ArgumentOptions.OT_NONE && (opt.Type & ArgumentOptions.OT_DEFAULT) == ArgumentOptions.OT_NONE) ||
                                    (argv[pnum][oplen] == '=' && opt.NumberOfParameters == 1 && (opt.Type & ArgumentOptions.OT_WE) != ArgumentOptions.OT_NONE) ||
                                    (argv[pnum][oplen] == ':' && opt.NumberOfParameters == 1 && (opt.Type & ArgumentOptions.OT_WC) != ArgumentOptions.OT_NONE))
                                {
                                    match = true;
                                }
                            }
                            catch (ArgumentOutOfRangeException) { }
                        }
                    }

                    if (match)
                    {
                        if (found_option != null)
                        {
                            WriteMsg.WriteLine("Error: Parameter {0} specified more than once.", WriteMsg.MsgLevel.Error, opt);
                            return false;
                        }
                        found_option = opt;
                    }
                }

                if (found_option != null)
                {
                    if (found_option.Value.NumberOfParameters == 0)
                    {
                        if (found_option.Value.Process != null && !found_option.Value.Process(found_option.Value, null))
                        {
                            WriteMsg.WriteLine("Error: Cannot process the '{0}' option", WriteMsg.MsgLevel.Error, found_option.Value.Name);
                            return false;
                        }
                    }
                    else if (found_option.Value.NumberOfParameters == 1)
                    {
                        if ((found_option.Value.Type & ArgumentOptions.OT_PREFIX) != ArgumentOptions.OT_NONE)
                        {
                            if (pnum + 1 >= argv.Length)
                            {
                                WriteMsg.WriteLine("Error: Cannot process the '{0}' option", WriteMsg.MsgLevel.Error, found_option.Value.Name);
                                return false;
                            }

                            if (!processOptionFile(options, optionPrefixes, unrecOption, found_option.Value, new[] { argv[pnum + 1] }))
                            {
                                WriteMsg.WriteLine("Error: Cannot process the '{0}' option", WriteMsg.MsgLevel.Error, found_option.Value.Name);
                                return false;
                            }
                            pnum++;
                        }
                        else if ((found_option.Value.Type & ArgumentOptions.OT_DEFAULT) != ArgumentOptions.OT_NONE)
                        {
                            bool needProcess = true;
                            int oplen = found_option.Value.Name.Length;
                            string prm = string.Empty;

                            if (argv[pnum][oplen] == '=')         //  -option=value (OT_WE)
                                prm = argv[pnum].Substring(oplen + 1, argv[pnum].Length - oplen - 1);
                            else if (argv[pnum][oplen] == ':')    //  -option:value (OT_WC)
                                prm = argv[pnum].Substring(oplen + 1, argv[pnum].Length - oplen - 1);
                            else
                            {

                                if (found_option.Value.DefaultProcess != null)
                                {
                                    defCallbacks.Add(found_option.Value);
                                    needProcess = false;
                                }
                                else
                                {
                                    WriteMsg.WriteLine("Error: The default value of '{0}' option is used but it does not have default value", WriteMsg.MsgLevel.Error, found_option.Value.Name);
                                    return false;
                                }
                                prm = argv[pnum + 1];
                                pnum++;
                            }
                            if (needProcess)
                            {
                                if (!processOptionFile(options, optionPrefixes, unrecOption, found_option.Value, new[] { prm }))
                                {
                                    WriteMsg.WriteLine("Error: Cannot process the '{0}' option", WriteMsg.MsgLevel.Error, found_option.Value.Name);
                                    return false;
                                }
                            }
                        }
                        else
                        {
                            int oplen = found_option.Value.Name.Length;
                            string prm = string.Empty;

                            if (argv[pnum][oplen] == '=')         //  -option=value (OT_WE)
                                prm = argv[pnum].Substring(oplen + 1, argv[pnum].Length - oplen - 1);
                            else if (argv[pnum][oplen] == ':')    //  -option:value (OT_WC)
                                prm = argv[pnum].Substring(oplen + 1, argv[pnum].Length - oplen - 1);
                            else if (argv[pnum][oplen] > 0)       //  -optionvalue (OT_WOS)
                                prm = argv[pnum].Substring(oplen, argv[pnum].Length - oplen);
                            else
                            {                                     // -option value (OT_WS)
                                if (pnum + 1 >= argv.Length)
                                {
                                    WriteMsg.WriteLine("Error: Cannot process the '{0}' option", WriteMsg.MsgLevel.Error, found_option.Value.Name);
                                    return false;
                                }
                                prm = argv[pnum + 1];
                                pnum++;
                            }

                            if (!processOptionFile(options, optionPrefixes, unrecOption, found_option.Value, new[] { prm }))
                            {
                                WriteMsg.WriteLine("Error: Cannot process the '{0}' option", WriteMsg.MsgLevel.Error, found_option.Value.Name);
                                return false;
                            }
                        }
                    }
                    else if (found_option.Value.NumberOfParameters > 1)
                    {
                        if (pnum + found_option.Value.NumberOfParameters >= argv.Length)
                        {
                            WriteMsg.WriteLine("Error: Cannot process the '{0}' option", WriteMsg.MsgLevel.Error, found_option.Value.Name);
                            return false;
                        }

                        if (found_option.Value.Process != null && !found_option.Value.Process(found_option.Value, argv[pnum + 1].Split(' ')))
                        {
                            WriteMsg.WriteLine("Error: Cannot process the '{0}' option", WriteMsg.MsgLevel.Error, found_option.Value.Name);
                            return false;
                        }

                        pnum += found_option.Value.NumberOfParameters;
                    }

                    if (found_option.Value.Needed > 0)
                        neededOptions.Remove(found_option.Value.Needed);
                }
                else
                {
                    if (Strchr(optionPrefixes, argv[pnum][0]) != null)
                    {
                        if (unrecOption != null)
                        {
                            unrecOption(argv[pnum]);
                        }
                        else
                        {
                            WriteMsg.WriteLine("Warning: Unrecognized parameter: '{0}'", WriteMsg.MsgLevel.Warning, argv[pnum]);
                        }
                    }
                }
            }

            foreach (var opt in defCallbacks)
            {
                opt.DefaultProcess(opt);
            }

            foreach (var pair in neededOptions)
            {
                if (pair.Value.Count > 1)
                    WriteMsg.WriteLine($"Error: One of the options {string.Join(", ", pair.Value)} should be provided.", WriteMsg.MsgLevel.Error);
                else
                    WriteMsg.WriteLine($"Error: Option {pair.Value.First()} is needed", WriteMsg.MsgLevel.Error);
            }

            return neededOptions.Count == 0;
        }

        public static void WriteHelp(Option[] options, bool displayInternal = false)
        {
            WriteMsg.WriteLine("", WriteMsg.MsgLevel.Silent);
            WriteMsg.WriteLine("  Options:", WriteMsg.MsgLevel.Silent);
            WriteMsg.WriteLine(Environment.NewLine, WriteMsg.MsgLevel.Silent);
            foreach (var opt in options)
            {
                if (opt.Internal && !displayInternal)
                    continue;                

                if (opt.Type == ArgumentOptions.OT_NONE)
                    WriteMsg.WriteLine("  {0}", WriteMsg.MsgLevel.Silent, opt.Name);
                else
                {
                    if ((opt.Type & ArgumentOptions.OT_WE) != 0 || opt.Type == ArgumentOptions.OT_DEFAULT)
                        WriteMsg.WriteLine("  {0}={1}", WriteMsg.MsgLevel.Silent, opt.Name, opt.ParameterName);
                    if ((opt.Type & ArgumentOptions.OT_WC) != 0 || opt.Type == ArgumentOptions.OT_DEFAULT)
                        WriteMsg.WriteLine("  {0}:{1}", WriteMsg.MsgLevel.Silent, opt.Name, opt.ParameterName);                    
                    if ((opt.Type & ArgumentOptions.OT_WOS) != 0)
                        WriteMsg.WriteLine("  {0}{1}", WriteMsg.MsgLevel.Silent, opt.Name, opt.ParameterName);
                    if ((opt.Type & ArgumentOptions.OT_WS) != 0)
                        WriteMsg.WriteLine("  {0} {1}", WriteMsg.MsgLevel.Silent, opt.Name, opt.ParameterName);
                }
                if (!string.IsNullOrEmpty(opt.Description))
                {
                    WriteMsg.WriteLine("    " + opt.Description, WriteMsg.MsgLevel.Silent);
                    WriteMsg.WriteLine("", WriteMsg.MsgLevel.Silent);
                }
            }
        }

        private static int? Strchr(string originalString, char charToSearch)
        {
            int? found = originalString.IndexOf(charToSearch);
            return found > -1 ? found : null;
        }

        private static bool processOptionFile(Option[] options, string optionPrefixes, DelegateOption unrecOption, Option found_option, string[] found_option_args)
        {
            if ((found_option.Type & ArgumentOptions.OT_OPTION_FILE) != ArgumentOptions.OT_NONE)
            {
                bool ret = true;
                if (found_option.Process != null)
                    return found_option.Process(found_option, found_option_args);

                if (ret)
                {
                    List<string> @params = new List<string>();
                    string param = ""; ;
                    int qmc = 0;
                    string[] lines = System.IO.File.ReadAllLines(found_option_args[0]);
                    foreach (var line in lines)
                    {
                        if (qmc % 2 != 0)
                            param += " ";
                        qmc += countQuotationMarks(line);
                        param += line;

                        if (qmc % 2 == 0)
                        {
                            @params.Add(removeQuotationMarks(param));

                            qmc = 0;
                            param = "";
                        }
                        if (qmc % 2 != 0)
                            @params.Add(removeQuotationMarks(param));
                    }

                    return ProcessArguments(@params.ToArray(), options, optionPrefixes, unrecOption);
                }
            }
            else
            {
                if (found_option.Process != null) return found_option.Process(found_option, found_option_args);
            }

            return true;
        }

        private static string removeQuotationMarks(string param)
        {
            throw new NotImplementedException();
        }

        private static int countQuotationMarks(string line)
        {
            throw new NotImplementedException();
        }
    }
}
